File: ios/Classes/BannerView/BannerMethodHandler.swift
//
//  BannerMethodHandler.swift
//  clever_ads_solutions
//
//  Copyright © 2024 CleverAdsSolutions LTD, CAS.AI. All rights reserved.
//

import CleverAdsSolutions
import Flutter
import Foundation

private let channelName = "cleveradssolutions/banner"

class BannerMethodHandler: AdMethodHandler<BannerView> {
    init(with registrar: FlutterPluginRegistrar, _ contentInfoHandler: AdContentInfoMethodHandler) {
        super.init(with: registrar, on: channelName, contentInfoHandler)
    }

    override func onMethodCall(_ instance: AdMethodHandler<BannerView>.Ad, _ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        switch call.method {
        case "getCASId": getCASId(instance.ad.banner, result)
        case "setCASId": setCASId(instance.ad.banner, call, result)
        case "setSize": setSize(instance.ad, call, result)
        case "isLoaded": isLoaded(instance.ad.banner, result)
        case "getContentInfo": getContentInfo(instance.contentInfoId, result)
        case "isAutoloadEnabled": isAutoloadEnabled(instance.ad.banner, result)
        case "setAutoloadEnabled": setAutoloadEnabled(instance.ad.banner, call, result)
        case "load": load(instance.ad.banner, result)
        case "getRefreshInterval": getRefreshInterval(instance.ad.banner, result)
        case "setRefreshInterval": setRefreshInterval(instance.ad.banner, call, result)
        case "disableAdRefresh": disableAdRefresh(instance.ad.banner, result)
        case "dispose": dispose(instance, result)
        default: super.onMethodCall(instance, call, result)
        }
    }

    private func getCASId(_ bannerView: CASBannerView, _ result: @escaping FlutterResult) {
        result(bannerView.casID)
    }

    private func setCASId(_ bannerView: CASBannerView, _ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("casId", result) { casId in
            bannerView.casID = casId
        }
    }

    private func setSize(_ bannerView: BannerView, _ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        guard let map = call.arguments as? [String: Any] else {
            result(call.errorArgNil("arguments"))
            return
        }
        let adSize = BannerMethodHandler.getAdSize(map, bannerView.frame)
        bannerView.banner.adSize = adSize

        result(nil)
    }

    private func isLoaded(_ bannerView: CASBannerView, _ result: @escaping FlutterResult) {
        result(bannerView.isAdLoaded)
    }

    private func getContentInfo(_ contentInfoId: String, _ result: @escaping FlutterResult) {
        result(contentInfoId)
    }

    private func isAutoloadEnabled(_ bannerView: CASBannerView, _ result: @escaping FlutterResult) {
        result(bannerView.isAutoloadEnabled)
    }

    private func setAutoloadEnabled(_ bannerView: CASBannerView, _ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("isEnabled", result) { isEnabled in
            bannerView.isAutoloadEnabled = isEnabled
        }
    }

    private func load(_ bannerView: CASBannerView, _ result: @escaping FlutterResult) {
        bannerView.loadAd()
        result(nil)
    }

    private func getRefreshInterval(_ bannerView: CASBannerView, _ result: @escaping FlutterResult) {
        result(bannerView.refreshInterval)
    }

    private func setRefreshInterval(_ bannerView: CASBannerView, _ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("interval", result) { interval in
            bannerView.refreshInterval = interval
        }
    }

    private func disableAdRefresh(_ bannerView: CASBannerView, _ result: @escaping FlutterResult) {
        bannerView.disableAdRefresh()
        result(nil)
    }

    private func dispose(_ ad: AdMethodHandler<BannerView>.Ad, _ result: @escaping FlutterResult) {
        destroy(ad: ad)
        ad.ad.banner.destroy()

        result(nil)
    }

    static func getAdSize(_ args: [String: Any?]?, _ frame: CGRect) -> CASSize {
        if let size = args?["size"] as? [String: Any?] {
            if size["isAdaptive"] as? Bool == true {
                if let maxWidth = size["maxWidthDpi"] as? Int,
                   maxWidth != 0 {
                    return CASSize.getAdaptiveBanner(forMaxWidth: CGFloat(maxWidth))
                } else {
                    return CASSize.getAdaptiveBanner(forMaxWidth: frame.width)
                }
            } else {
                let width = size["width"] as? CGFloat ?? 0
                let mode = size["mode"] as? CGFloat ?? 0

                switch mode {
                case 2:
                    return CASSize.getAdaptiveBanner(forMaxWidth: width)
                case 3:
                    let height = size["height"] as? CGFloat ?? 0
                    return CASSize.getInlineBanner(width: width, maxHeight: height)
                default: switch width {
                    case 300: return CASSize.mediumRectangle
                    case 728: return CASSize.leaderboard
                    default: return CASSize.banner
                    }
                }
            }
        }
        return CASSize.banner
    }
}

File: android/src/main/kotlin/com/cleveradssolutions/plugin/flutter/bannerview/BannerMethodHandler.kt
package com.cleveradssolutions.plugin.flutter.bannerview

import android.content.Context
import com.cleveradssolutions.plugin.flutter.bridge.base.AdMethodHandler
import com.cleveradssolutions.plugin.flutter.sdk.AdContentInfoHandler
import com.cleveradssolutions.plugin.flutter.util.errorArgNull
import com.cleveradssolutions.plugin.flutter.util.getArgAndReturn
import com.cleveradssolutions.plugin.flutter.util.success
import com.cleversolutions.ads.AdSize
import com.cleversolutions.ads.android.CASBannerView
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

private const val CHANNEL_NAME = "cleveradssolutions/banner"

class BannerMethodHandler(
    binding: FlutterPluginBinding,
    contentInfoHandler: AdContentInfoHandler
) : AdMethodHandler<BannerView>(binding, CHANNEL_NAME, contentInfoHandler) {

    override fun onMethodCall(
        instance: Ad<BannerView>,
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        when (call.method) {
            "getCASId" -> getCASId(instance.ad.view, result)
            "setCASId" -> setCASId(instance.ad.view, call, result)
            "setSize" -> setSize(instance.ad.view, call, result)
            "isLoaded" -> isLoaded(instance.ad.view, result)
            "getContentInfo" -> getContentInfo(instance.contentInfoId, result)
            "isAutoloadEnabled" -> isAutoloadEnabled(instance.ad.view, result)
            "setAutoloadEnabled" -> setAutoloadEnabled(instance.ad.view, call, result)
            "load" -> load(instance.ad.view, result)
            "getRefreshInterval" -> getRefreshInterval(instance.ad.view, result)
            "setRefreshInterval" -> setRefreshInterval(instance.ad.view, call, result)
            "disableAdRefresh" -> disableAdRefresh(instance.ad.view, result)
            "dispose" -> dispose(instance, result)
            else -> super.onMethodCall(instance, call, result)
        }
    }

    private fun getCASId(bannerView: CASBannerView, result: MethodChannel.Result) {
        result.success(bannerView.casId)
    }

    private fun setCASId(
        bannerView: CASBannerView,
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        call.getArgAndReturn<String>("casId", result) {
            bannerView.casId = it
        }
    }

    private fun setSize(
        bannerView: CASBannerView,
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        val map = call.arguments as? Map<*, *> ?: return result.errorArgNull(call, "arguments")
        val adSize = parseSize(bannerView.context, map)
        bannerView.size = adSize

        result.success()
    }

    private fun isLoaded(bannerView: CASBannerView, result: MethodChannel.Result) {
        result.success(bannerView.isLoaded)
    }

    private fun getContentInfo(contentInfoId: String, result: MethodChannel.Result) {
        result.success(contentInfoId)
    }

    private fun isAutoloadEnabled(bannerView: CASBannerView, result: MethodChannel.Result) {
        result.success(bannerView.isAutoloadEnabled)
    }

    private fun setAutoloadEnabled(
        bannerView: CASBannerView,
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        call.getArgAndReturn<Boolean>("isEnabled", result) {
            bannerView.isAutoloadEnabled = it
        }
    }

    private fun load(bannerView: CASBannerView, result: MethodChannel.Result) {
        bannerView.load()
        result.success()
    }

    private fun getRefreshInterval(bannerView: CASBannerView, result: MethodChannel.Result) {
        result.success(bannerView.refreshInterval)
    }

    private fun setRefreshInterval(
        bannerView: CASBannerView,
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        call.getArgAndReturn<Int>("interval", result) {
            bannerView.refreshInterval = it
        }
    }

    private fun disableAdRefresh(bannerView: CASBannerView, result: MethodChannel.Result) {
        bannerView.disableAdRefresh()
        result.success()
    }

    private fun dispose(ad: Ad<BannerView>, result: MethodChannel.Result) {
        destroy(ad)
        ad.ad.view.destroy()

        result.success()
    }

    companion object {
        fun parseSize(context: Context, size: Map<*, *>): AdSize {
            if (size["isAdaptive"] == true) {
                return (size["maxWidthDpi"] as? Int).let {
                    if (it == null || it == 0) {
                        AdSize.getAdaptiveBannerInScreen(context)
                    } else {
                        AdSize.getAdaptiveBanner(context, it)
                    }
                }
            } else {
                val width = size["width"] as Int
                val height = size["height"] as Int
                val mode = size["mode"] as Int

                return when (mode) {
                    2 -> AdSize.getAdaptiveBanner(context, width)
                    3 -> AdSize.getInlineBanner(width, height)
                    else -> when (width) {
                        300 -> AdSize.MEDIUM_RECTANGLE
                        728 -> AdSize.LEADERBOARD
                        else -> AdSize.BANNER
                    }
                }
            }
        }
    }

}